### PR TITLE
[Rchain 3049]: expose RevAddress in Rholang

### PR DESCRIPTION
**Pull request by odeac**
## Overview
expose RevAddress validation as a system process
refactor SystemProcesses in order to remove the duplication

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3049

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by odeac** [comment]
bors try

**Comment by KentShikama** [comment]
@odeac Please merge from upstream/dev

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 1]
```diff
@@ -105,6 +107,36 @@ object SystemProcesses {
           } yield ()
       }
 
+      def getErrorMessagePar(str: String) =
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: Please add return type, and do so for all the functions you define from now on.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 2]
```diff
@@ -105,6 +107,36 @@ object SystemProcesses {
...
+        RevAddress
+          .parse(str)
+          .swap
+          .toOption
+          .fold(Par.defaultInstance)(GString(_))
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: let's not use `deafaultInstance` and just `Par()` instaed. The former is somewhat protobuf-specific

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: weren't we supposed to return an "Either"? Guess you do that in latter commits?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 3]
```diff
@@ -105,6 +107,36 @@ object SystemProcesses {
...
+            sequenceNumber
+            ) =>
+          for {
+            result <- F.delay(getErrorMessagePar(address))
+            produceResult <- space.produce(
+                              ack,
+                              ListParWithRandom(Seq(result), rand),
+                              persist = false,
+                              sequenceNumber
+                            )(UNLIMITED_MATCH_PHLO)
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: OMG we need to make system processes aware of the cost account. Not in this PR ofc.

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala [hunk 4]
```diff
@@ -30,9 +30,9 @@ object RhoSpec {
         } ++ Seq(
         SystemProcess.Definition[Task](
           "rho:io:stdlog",
-          Runtime.byteName(26),
+          Runtime.byteName(27),
           2,
-          26L,
+          27L,
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: I think those numbers should've changed in the previous commit?
What happened before you updated those numbers?
It'd be nice to have a registration mechanism that doesn't require magic constants synced throughout the whole codebase

File: casper/src/test/rholang/RevAddressTest.rho [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    @RhoSpec!("testSuite", *setup,
+      [
+        ("Validating a valid address returns Nil", *test_valid_address),
+        ("Validating a invalid address returns a non-Nil value", *test_invalid_address),
+      ])
+  } |
+
+  contract setup(returnCh) = {
+    returnCh!([])
+  } |
```
Comment (ArturGajowy) [inline casper/src/test/rholang/RevAddressTest.rho]: would be great if there was a way to skip defining the `setup` contract

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 6]
```diff
@@ -39,6 +39,23 @@ trait SystemProcesses[F[_]] {
   def validateRevAddress: (Seq[ListParWithRandomAndPhlos], Int) => F[Unit]
 }
 
+object RhoType {
+  object ByteArray {
+    import coop.rchain.models.rholang.implicits._
+    def unapply(p: Par): Option[Array[Byte]] =
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: :100:

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 7]
```diff
@@ -124,12 +124,12 @@ object SystemProcesses {
...
       }
 
-      def getErrorMessagePar(str: String) =
+      private def getErrorMessagePar(str: String): Par =
         RevAddress
           .parse(str)
           .swap
           .toOption
-          .fold(Par.defaultInstance)(GString(_))
+          .fold(Par.defaultInstance)(GString)
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: Please fixup the relevant commit with the `fix compilation warnings in SystemProcesses.scala` one

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 8]
```diff
@@ -68,59 +68,85 @@ object SystemProcesses {
...
+                  produceResult <- space.produce(
+                                    ch,
+                                    ListParWithRandom(values, rand),
+                                    persist = false,
+                                    sequenceNumber
+                                  )(UNLIMITED_MATCH_PHLO)
+                  _ <- foldResult(produceResult)
+                } yield ()
+
+              Some((produce, args))
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: It's very strange for an extractor to always return `Some`. I guess we should be returning `case _ => None`

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 9]
```diff
@@ -68,59 +68,85 @@ object SystemProcesses {
...
 
-      private def foldResult(
-          produceResult: Either[InterpreterError, Option[(ContWithMetaData, Channels)]]
-      ): F[Unit] =
-        produceResult.fold(
-          _ => F.raiseError(OutOfPhlogistonsError),
-          _.fold(F.unit) {
-            case (cont, channels) =>
-              dispatcher.dispatch(unpackCont(cont), channels.map(_.value), cont.sequenceNumber)
+      object ContractCall {
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: this object is very hard to read. I guess moving the definitions past their usages would help somewhat. Also, the nested `def produce` is quite surprising. I think a `val producer: Producer = (values: Seq[Par], channel: Par)) => ...` would be better

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 10]
```diff
@@ -68,59 +68,85 @@ object SystemProcesses {
...
-                              sequenceNumber
-                            )(UNLIMITED_MATCH_PHLO)
-            _ <- foldResult(produceResult)
+            _ <- produce(Seq(Par.defaultInstance), ack)
           } yield ()
       }
 
       def stdErr: (Seq[ListParWithRandomAndPhlos], Int) => F[Unit] = {
-        case (Seq(ListParWithRandomAndPhlos(Seq(arg), _, _)), _) =>
+        case ContractCall(_, Seq(arg)) =>
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: thx for the deduplication! That's a huge improvement!

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 11]
```diff
@@ -46,13 +46,24 @@ object RhoType {
...
   }
 
   object String {
     def unapply(p: Par): Option[String] =
       p.singleExpr().collect {
         case Expr(GString(bs)) => bs
       }
+
+    def apply(s : String) : Par = GString(s)
+    def apply(maybeStr : Option[String]) : Par =
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: please get rid of this - I'd rather have this Option handled explicitly

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 12]
```diff
@@ -86,16 +86,23 @@ object SystemProcesses {
...
-        ): F[Unit] =
-          produceResult.fold(
-            _ => F.raiseError(OutOfPhlogistonsError),
-            _.fold(F.unit) {
-              case (cont, channels) =>
-                dispatcher.dispatch(unpackCont(cont), channels.map(_.value), cont.sequenceNumber)
-            }
-          )
+        private def produce(rand : Blake2b512Random, sequenceNumber : Int)
+                           (values: Seq[Par], ch: Par): F[Unit] =
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: Let's not glue this all together, as it's now unclear which parameter is used when. The second parameter list should be a part of an inner function.

I'd drop this commit and start again, applying the comments related to ContractCall object

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala (hunk 1 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala (hunk 2 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: The Either makes sense for a parse function. The validate returns an error message if there was a problem or a Nil if the validation was successful

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala (hunk 4 above)
Comment (odeac) [reply casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: done

File: casper/src/test/rholang/RevAddressTest.rho (hunk 5 above)
Comment (odeac) [reply casper/src/test/rholang/RevAddressTest.rho]: I'll do that in a future PR

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala (hunk 9 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: This is cleaned up a little bit in the next commit

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala (hunk 8 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: fixed

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala (hunk 11 above)
Comment (odeac) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: True.

It seemed like a good idea at the time.

**Pull request merged**